### PR TITLE
One single flag that determines if we are in inference

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/models/gpt/gpt_model.py [hunk 1]
```diff
@@ -308,7 +308,7 @@ def _preprocess(
         # If decoder_input is provided (not None), then input_ids and position_ids are ignored.
         # Otherwise, apply embedding layer on input_ids and position_ids to get decoder_input.
 
-        in_inference_mode = inference_context is not None and not self.training
+        in_inference_mode = inference_context is not None and inference_context.is_active
```
Comment (sidsingh-nvidia) [inline megatron/core/models/gpt/gpt_model.py]: should we change this to a more strict assert?
```
if inference_context.is_active:
     assert inference_context is not None
```

**Review by sidsingh-nvidia — APPROVED** [review]

File: megatron/core/inference/utils.py [hunk 2]
```diff
@@ -80,6 +105,8 @@ def _init_moe_expert_cache(model):
     """
     Initialize the cache of MoE layers once
     """
+    from megatron.core.transformer.moe.moe_layer import MoELayer
```
Comment (sidsingh-nvidia) [inline megatron/core/inference/utils.py:119]: This can be removed, right?

**Review by sidsingh-nvidia — CHANGES_REQUESTED** [review]

File: megatron/core/inference/utils.py (hunk 2 above)
Comment (tdene) [reply megatron/core/inference/utils.py:119]: The `def walk` further down still uses this:

```
    def walk(module):
        # Collect from MoELayer fields
        if isinstance(module, MoELayer):
            oid = id(module)
            if oid not in seen_modules:
                moe_layer_cache.append(module)

        for child in module.children():
            walk(child)
```

The reason why this was removed from the top of the file and re-added in this helper is to prevent import-chain loops. The transformer layers import this `utils.py` file now, so we can't safely import from the transformer layers at the top of the file.

It's just this one method that has this issue.

**Review by sidsingh-nvidia — APPROVED** [review]

File: megatron/core/inference/engines/dynamic_engine.py
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py:1]: Is setting `InferenceMode.set_active()` only in `DynamicEngine` sufficient for the context / text gen controller unit tests which don't instantiate an engine?

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:1]: Thanks for the heads-up. I just made a commit that should be a blanket fix to all unit tests.

Waiting for CI to see if that was sufficient.

Comment (santhnm2) [reply megatron/core/inference/engines/dynamic_engine.py:1]: Small style suggestion: could we add a `contextmanager` to `InferenceMode` so that we can do `with inference_mode():` wrappers instead of the `try / finally` pattern? I guess the `contextmanager` would just do the `try / finally` internally.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:1]: Done

**Review by yaox12 — APPROVED** [review]
LGTM for the MoE part.

**Comment by tdene** [comment]
@santhnm2 the unit test dimensions change from `max_seq_len` to `1` because this PR fixes a pre-existing bug on `main`.

Static inference hardcodes `materialize_only_last_token_logits` to `True`. On `main`, unit tests ignore this by running inference without being in inference mode. This PR makes ALL inference run in inference mode, including unit tests which previously ran inference outside of inference mode, so they ignored `materialize_only_last_token_logits`.

**Comment by santhnm2** [comment]
> @santhnm2 the unit test dimensions change from `max_seq_len` to `1` because this PR fixes a pre-existing bug on `main`.
> 
> Static inference hardcodes `materialize_only_last_token_logits` to `True`. On `main`, unit tests ignore this by running inference without being in inference mode. This PR makes ALL inference run in inference mode, including unit tests which previously ran inference outside of inference mode, so they ignored `materialize_only_last_token_logits`.

I see, so were there failing tests that were expecting `max_seq_len` instead of 1?

**Comment by tdene** [comment]
> > @santhnm2 the unit test dimensions change from `max_seq_len` to `1` because this PR fixes a pre-existing bug on `main`.
> > Static inference hardcodes `materialize_only_last_token_logits` to `True`. On `main`, unit tests ignore this by running inference without being in inference mode. This PR makes ALL inference run in inference mode, including unit tests which previously ran inference outside of inference mode, so they ignored `materialize_only_last_token_logits`.
> 
> I see, so were there failing tests that were expecting `max_seq_len` instead of 1?

You be the judge of what the correct solution was.

On `main`, there were working tests that expected `max_seq_len` instead of 1. They only worked because they were not actually entering inference mode, so they were never checking the `materialize_only_last_token_logits`.

But static inference was intentionally hardcoding `materialize_only_last_token_logits` to `True`, so from that point of view those tests were flawed.

I can go back and change static inference to stop hardcoding `materialize_only_last_token_logits` to `True`, pass it as `False` into these tests, and then the tests would get `max_seq_len` and we wouldn't have to change their golden asserts. But I'm not sure that is worth doing since static inference is getting deprecated?

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by yaoyu-33 — APPROVED** [review]
